src/main/java/com/impactupgrade/integration/sfdc/SFDCPartnerAPIClient.java

```diff
@@ -670,43 +671,46 @@ public Map<String, String> toMapOfStrings(SObject sObject) {
     }));
   }
 
-  private QueryResult _query(int count, String queryString) throws ConnectionException, InterruptedException {
-    try {
-      return partnerConnection.get().query(queryString);
-    } catch (ApiFault e) {
-      log.error("query failed due to {}: {}", e.getExceptionCode(), e.getExceptionMessage(), e);
-      throw e;
-    } catch (ConnectionException e) {
-      log.warn("query attempt {} failed due to connection issues; retrying in 5s", count, e);
-      Thread.sleep(5000);
-
-      if (count == 5) {
-        log.error("unable to complete query by attempt {}", count);
-        // rethrow exception, since the whole flow simply needs to halt at this point
-        throw e;
-      }
+  private QueryResult _query(String queryString) throws ConnectionException, InterruptedException {
+    return queryWithRetry(() -> partnerConnection.get().query(queryString));
+  }
 
-      return _query(count + 1, queryString);
-    }
+  private QueryResult _queryMore(String previousQueryLocator) throws ConnectionException, InterruptedException {
+    return queryWithRetry(() -> partnerConnection.get().queryMore(previousQueryLocator));
   }
 
-  private QueryResult _queryMore(int count, String previousQueryLocator) throws ConnectionException, InterruptedException {
-    try {
-      return partnerConnection.get().queryMore(previousQueryLocator);
-    } catch (ApiFault e) {
-      log.error("query failed due to {}: {}", e.getExceptionCode(), e.getExceptionMessage(), e);
-      throw e;
-    } catch (ConnectionException e) {
-      log.warn("query attempt {} failed due to connection issues; retrying in 5s", count, e);
-      Thread.sleep(5000);
+  private QueryResult queryWithRetry(CallableQuery<QueryResult> callableQuery) throws ConnectionException, InterruptedException {
+    int count = 0;
+    int maxTries = 5;
+    while(true) {
+      try {
+        return callableQuery.call();
+      } catch (Exception e) {
+        // check if max tries exceeded
+        if (++count == maxTries) {
+          log.error("unable to complete query by attempt {}", count);
+          // rethrow exception, since the whole flow simply needs to halt at this point
+          throw e;
+        }
 
-      if (count == 5) {
-        log.error("unable to complete query by attempt {}", count);
-        // rethrow exception, since the whole flow simply needs to halt at this point
-        throw e;
-      }
+        if (e instanceof ApiFault) {
```

Comment:
Shouldn't this line go up above the maxRetries check? ApiFaults do not get retried

Reply:
Ah, I get the point about doing a retry after the login (nice idea!). Just thinking through how to avoid doing retries for the other ApiFaults